crds/main.go

```diff
@@ -0,0 +1,374 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/pulumi/pulumi/pkg/v2/codegen/schema"
+	pschema "github.com/pulumi/pulumi/pkg/v2/codegen/schema"
+	unstruct "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const tool = "crdgen"
+
+func main() {
+	if len(os.Args) < 2 {
+		fmt.Fprintf(os.Stderr, "error: expected <resourcedefinition.yaml> argument\n")
+		os.Exit(-1)
+	}
+
+	yamlPath := os.Args[1]
+	yamlFile, err := ioutil.ReadFile(yamlPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error reading the yaml file: %v\n", err)
+		os.Exit(-1)
+	}
+
+	crd, err := UnmarshalYaml(yamlFile)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error unmarshalling yaml: %v\n", err)
+		os.Exit(-1)
+	}
+
+	types := getTypes(crd)
+
+	// Generate TypeScript for the types.
+	code, err := generateTypeScriptTypes(types)
+	if err != nil {
+		log.Printf("error: %v", err)
+	}
+
+	// Write out the generated code to standard output.
+	fmt.Println(code)
+}
+
+// getTypes parses all the types within the given crd (an unmarshalled
+// resourcedefinition.yaml file). Returns a mapping from each type's name in the
+// format "{.spec.group}/{.spec.versions[*].name}:{.spec.names.kind}" to its
+// corresponding pschema.ObjectTypeSpec.
+func getTypes(crd unstruct.Unstructured) map[string]pschema.ObjectTypeSpec {
+	versions, _, _ := NestedMapSlice(crd.Object, "spec", "versions")
+	metadataName, _, _ := unstruct.NestedString(crd.Object, "metadata", "name")
+	metadataName = cleanMetadataName(metadataName)
+	kind, _, _ := unstruct.NestedString(crd.Object, "spec", "names", "kind")
+
+	types := make(map[string]pschema.ObjectTypeSpec)
+	for _, version := range versions {
+		schema, _, _ := unstruct.NestedMap(version, "schema", "openAPIV3Schema")
+		versionName, _, _ := unstruct.NestedString(version, "name")
+		baseRef := fmt.Sprintf("crds:%s/%s:%s", metadataName, versionName, kind)
+		addType(schema, baseRef, types)
+	}
+
+	return types
+}
+
+/*
+getType converts the given OpenAPI v3 `schema` to a ObjectTypeSpec and adds
+it to the `types` map under the given `name`. Recursively converts and adds all
+nested schemas as well.
+*/
+func addType(schema map[string]interface{}, name string, types map[string]pschema.ObjectTypeSpec) {
+	description, _, _ := unstruct.NestedString(schema, "description")
+	schemaType, _, _ := unstruct.NestedString(schema, "type")
+	properties, _, _ := unstruct.NestedMap(schema, "properties")
+	required, _, _ := unstruct.NestedStringSlice(schema, "required")
+
+	propertySpecs := make(map[string]pschema.PropertySpec)
```

Comment:
FYI, you can write this like `map[string]pschema.PropertySpec{}` to save a little typing over explicitly calling `make`. Mostly a matter of style.